src/Analyser/NodeScopeResolver.php

```diff
@@ -3314,11 +3316,22 @@ private function processAssignVar(
 				if ($propertyReflection->canChangeTypeAfterAssignment()) {
 					$scope = $scope->assignExpression($var, $assignedExprType);
 				}
+				if (!$propertyReflection->getWritableType()->isSuperTypeOf($assignedExprType)->yes()) {
+					$throwPoints[] = ThrowPoint::createExplicit($scope, new ObjectType(TypeError::class), $assignedExpr, false);
+				}
 			} else {
 				// fallback
 				$assignedExprType = $scope->getType($assignedExpr);
 				$nodeCallback(new PropertyAssignNode($var, $assignedExpr, $isAssignOp), $scope);
 				$scope = $scope->assignExpression($var, $assignedExprType);
+				// simulate dynamic property assign to get throw points
+				$throwPoints = array_merge($throwPoints, $this->processExprNode(
+					new MethodCall($var->var, '__set'),
```

Comment:
This needs some tests :)

1) Class without the `__set` magic method - does not throw anything (dead catch reported).
2) Class with `__set` magic method and `@throws void` annotation (dead catch reported).
3) Class with `__set` magic method and `@throws SomeException` (dead catch not reported).

Reply:
:+1: 

For the first case, there is a excplitThrowPoint already, when method reflection is not found.
I'm not sure with the purpose of this throw point, so I'm gonna check it.
https://github.com/phpstan/phpstan-src/blob/21674384d26e6680d98545db11a6dfa0e12eb4fb/src/Analyser/NodeScopeResolver.php#L1994

Reply:
It was from this case
https://github.com/phpstan/phpstan-src/commit/84b09349dc04fa30097900b5deb8e316b1657e4b

Reply:
Ok, understood
I shouldn’t process `__set` if its not declared.